lib/service/cfg.go

```diff
@@ -276,6 +276,26 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
 	return false, nil
 }
 
+// ApplyCAPins assigns the given CA pin(s), but only if each pin is not empty.
+func (cfg *Config) ApplyCAPins(caPins []string) error {
```

Comment:
This is a good point, and on further thinking I realized that if the file itself is empty, that file content would not be "filtered out" as-is. It was the case before that we ignored a single empty ca_pin, but not multiple. I'm not sure why. I'll have to check the prior behavior with multiple empty strings.

Reply:
My initial reaction is that an empty file is likely a config error, so I'd be tempted to fail in that case, but that also changes past behavior. :/

Let us know what you found out on this one.

Reply:
So when adding a node with prior behavior, I observed the following behavior:

1. no ca_pin in config: joining node prints a warning about man-in-the-middle attacks without using ca_pin, but it still starts
2. single blank ca_pin: same warning, still starts (as expected since we skip it if there's just one ca_pin entry and it's blank)
3. multiple blank ca_pin entries: error message when starting the node:

```
2022-05-23T16:09:07-07:00 ERRO [PROC:1]    Node failed to establish connection to cluster: invalid format for certificate pin, expected algorithm:pin, object not found. service/connect.go:86
```

Whether we should preserve this behavior, I'm not sure. If we do, then I could at least add a warning or something about blank CA pins. The message could be more helpful at least, if we do keep it like this.

Reply:
So you're saying if we find empty config file values, just log it as a warning and preserve the behavior that the error occurs down the line when it gets the empty CAPin (`ERRO [PROC:1]...`)

Reply:
I'm suggesting:

1. Let's make sure ApplyCAPins keeps the behavior of the old `if len > 1`, preferably erroring if any pins are empty in that scenario
2. Let's check the return from TryReadValueAsFile in ApplyCAPins to make sure it's not empty and error if it is.

Reply:
1. agreed
2. agreed, but I think we should check for empty file contents in TryReadValueAsFile and log a warning as well. This will provide some hint of what went wrong parsing for any code that goes through this function. It may also hint at a config mistake, even if there is no error thrown,  for example I tested `auth_service.tokens` parsing behavior on master: empty tokens or empty token files show no warning nor error and proceed to start the server. Same for an empty `teleport.auth_token` file:

```
teleport:
  # valid, no warning nor error
  auth_token: /Users/gavin/some-empty-file
...
auth_service:
  tokens:
    # both valid, no warning/error
    - node:/Users/gavin/some-empty-file
    - "node:"
```

3. Take @nklaassen's suggestion to allow multiple values in a file. I can't foresee this breaking any past behavior.

Reply:
Partly agree with 2., see other comments.

I don't see 3. implemented?

Reply:
I didn't implement 3. yet, as I wasn't sure if everyone was on board with that

Reply:
3. Implemented and I updated tests for config value filepath in tokens/pins